components/script/dom/htmlselectelement.rs

```diff
@@ -185,49 +190,60 @@ impl HTMLSelectElementMethods for HTMLSelectElement {
         self.upcast::<HTMLElement>().labels()
     }
 
-    // https://html.spec.whatwg.org/multipage/#dom-select-length
-    fn SetLength(&self, value: u32) {
-        let length  = self.Length();
-        let node = self.upcast::<Node>();
-        if value < length {     // truncate the number of option elements
-            let mut iter = node.rev_children().take((length - value) as usize);
-            while let Some(child) = iter.next() {
-                if let Err(e) = node.RemoveChild(&child) {
-                    warn!("Error removing child of HTMLSelectElement: {:?}", e);
-                }
-            }
-        } else if value > length {  // add new blank option elements
-            let document = document_from_node(self);
-            for _ in 0..(value - length) {
-                let element = HTMLOptionElement::new(atom!("option"), None, &document.upcast());
-                if let Err(e) = node.AppendChild(element.upcast()) {
-                    warn!("error appending child of HTMLSelectElement: {:?}", e);
-                }
+    // https://html.spec.whatwg.org/multipage/#dom-select-options
+    fn Options(&self) -> Root<HTMLOptionsCollection> {
+        if let Some(options) = self.options.get() {
+            return options;
+        }
+
+        #[derive(JSTraceable, HeapSizeOf)]
+        struct OptionsFilter;
+        impl CollectionFilter for OptionsFilter {
+            fn filter<'a>(&self, elem: &'a Element, _root: &'a Node) -> bool {
+                elem.is::<HTMLOptionElement>()
             }
         }
-    }
+        let filter = box OptionsFilter;
+        let window = window_from_node(self);
+        let options = HTMLOptionsCollection::new(window.r(), self.upcast(), filter);
+        self.options.set(Some(&options));
+        options
+      }
 
     // https://html.spec.whatwg.org/multipage/#dom-select-length
     fn Length(&self) -> u32 {
-        self.upcast::<Node>()
-            .traverse_preorder()
-            .filter_map(Root::downcast::<HTMLOptionElement>)
-            .count() as u32
+        self.Options().Length() as u32
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-select-length
+    fn SetLength(&self, length: u32) {
+        self.Options().SetLength(length)
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-select-item
     fn Item(&self, index: u32) -> Option<Root<Element>> {
-        self.upcast::<Node>()
-            .traverse_preorder()
-            .filter_map(Root::downcast::<HTMLOptionElement>)
-            .nth(index as usize)
-            .map(|item| Root::from_ref(item.upcast()))
+        self.Options().IndexedGetter(index)
```

Comment:
Same as the previous similar - there is no such method (Item) for HTMLOptionsCollection.